neural_lam/weather_dataset.py

```diff
@@ -1,262 +1,373 @@
-        sample = full_sample[
-            subsample_index : subsample_end_index : self.subsample_step
-        ]
-        # (N_t, dim_y, dim_x, d_features')
-
-        # Remove feature 15, "z_height_above_ground"
-        sample = torch.cat(
-            (sample[:, :, :, :15], sample[:, :, :, 16:]), dim=3
-        )  # (N_t, dim_y, dim_x, d_features)
-
-        # Accumulate solar radiation instead of just subsampling
-        rad_features = full_sample[:, :, :, 2:4]  # (N_t', dim_y, dim_x, 2)
-        # Accumulate for first time step
-        init_accum_rad = torch.sum(
-            rad_features[: (subsample_index + 1)], dim=0, keepdim=True
-        )  # (1, dim_y, dim_x, 2)
-        # Accumulate for rest of subsampled sequence
-        in_subsample_len = (
-            subsample_end_index - self.subsample_step + subsample_index + 1
-        )
-        rad_features_in_subsample = rad_features[
-            (subsample_index + 1) : in_subsample_len
-        ]  # (N_t*, dim_y, dim_x, 2), N_t* = (N_t-1)*ss_step
-        _, dim_y, dim_x, _ = sample.shape
-        rest_accum_rad = torch.sum(
-            rad_features_in_subsample.view(
-                self.original_sample_length - 1,
-                self.subsample_step,
-                dim_y,
-                dim_x,
-                2,
-            ),
-            dim=1,
-        )  # (N_t-1, dim_y, dim_x, 2)
-        accum_rad = torch.cat(
-            (init_accum_rad, rest_accum_rad), dim=0
-        )  # (N_t, dim_y, dim_x, 2)
-        # Replace in sample
-        sample[:, :, :, 2:4] = accum_rad
-
-        # Flatten spatial dim
-        sample = sample.flatten(1, 2)  # (N_t, N_grid, d_features)
-
-        # Uniformly sample time id to start sample from
-        init_id = torch.randint(
-            0, 1 + self.original_sample_length - self.sample_length, ()
+            da_state = self.da_state
+
+        if self.da_forcing is not None:
+            if "ensemble_member" in self.da_forcing.dims:
+                raise NotImplementedError(
+                    "Ensemble member not yet supported for forcing data"
+                )
+            da_forcing = self.da_forcing
+        else:
+            da_forcing = None
+
+        # handle time sampling in a way that is compatible with both analysis
+        # and forecast data
+        da_state = self._sample_time(
+            da=da_state, idx=idx, n_steps=2 + self.ar_steps
         )
-        sample = sample[init_id : (init_id + self.sample_length)]
-        # (sample_length, N_grid, d_features)
+
+        if da_forcing is not None:
+            das_forcing = []
+            for n in range(self.forcing_window_size):
+                da_ = self._sample_time(
+                    da=da_forcing,
+                    idx=idx,
+                    n_steps=self.ar_steps,
+                    n_timesteps_offset=n,
+                )
+                if n > 0:
+                    da_ = da_.drop_vars("time")
+                das_forcing.append(da_)
```

Comment:
If I understand this code correctly this actually changes the forcing window behavior to put the "window" after the current time. We should probably stick to what is done in the literature as the default: 3-step windowing of forcing means including forcing from time steps (t-1, t, t+1) as inputs at time t. I think this code includes forcing for times (t, t+1, t+2).

It would be logical to me that `forcing_window_size=5` then would mean including forcing from (t-2,t-1,t,t+1,t+2) at time t. However, it is unclear what an even `forcing_window_size` then means. One could just set a convention, like including additional forcing times in the future first (`forcing_window_size=4` being (t-1, t, t+1, t+2)). Another option would be to decompose forcing window size into past and future, e.g. something like `--include_past_forcing 1 --include_future_forcing 1` resulting in a (t-1,t,t+1) window.

Reply:
I have implemented the `include_past_forcing 1 --include_future_forcing 1` variant. Because it would be hard to argue whether the past or the future should be considered more importent (philosofical, I now). There is one issue left here. The way the `forecast` type data is handled currently is like this:

_# We for now
  # simply select a analysis time and the first `n_steps` forecast
  # times (given no offset). Note that this means that we get one
  # sample per forecast, always starting at forecast time 2._

From this follows, that there is now past that could be retrieved for `forecast` type data. So currently I implemented that the window starts at t0 and spans into the future u to `include_future_forcing + 1`.
@joeloskarsson Is there a better way to do this? The way forecast data is handled right now seems really wasteful wrt training data. How much do we care about `forecast` data (are we all training on reanalysis soon anywas?)?



```

    def _slice_forcing_time(self, da, idx):
        """
        Slice the forcing data to include past and future time steps.

        Parameters
        ----------
        da : xr.DataArray
            The dataarray to slice. This is expected to have a `time`
            dimension if the datastore is providing analysis only data, and a
            `analysis_time` and `elapsed_forecast_duration` dimensions if the
            datastore is providing forecast data.
        idx : int
            The index of the time step to start the sample from.

        Returns
        -------
        da : xr.DataArray
            The sliced forcing data.
        """
        if self.datastore.is_forecast:
            # Handle forecast data with `analysis_time` and `elapsed_forecast_duration`
            da = da.isel(
                analysis_time=idx,
                elapsed_forecast_duration=slice(0, self.include_future_forcing + 1),
            )
            # Create a new time dimension
            da["time"] = da.analysis_time + da.elapsed_forecast_duration
            da = da.swap_dims({"elapsed_forecast_duration": "time"})
        else:
            # Handle analysis only data with `time` dimension
            start_idx = idx - self.include_past_forcing
            end_idx = idx + self.include_future_forcing + 1
            da = da.isel(time=slice(start_idx, end_idx))

        return da
```

Reply:
Wait I found the solution, haha:
```

    def _slice_time(self, da, idx, n_steps: int, n_timesteps_offset: int = 0):
        """
        Produce a time slice of the given dataarray `da` (state or forcing)
        starting at `idx` and with `n_steps` steps. The `n_timesteps_offset`
        parameter is used to offset the start of the slice, for example to
        exclude the first two steps when slicing the forcing data (and to
        produce the windowing indices of forcing data by increasing the offset
        for each window).
```

Reply:
I guess you figured this out? At least for the MEPS case there used to be past forcing from the time steps of the two initial conditions that could be used. 

But in general we don't need to tweak things too much initially to accommodate forecast datasets.